addons/binding/org.openhab.binding.kodi/src/main/java/org/openhab/binding/kodi/internal/protocol/KodiConnection.java

```diff
@@ -568,65 +568,81 @@ public synchronized void playURI(String uri) {
-                    return channelGroup.get("channelgroupid").getAsInt();
-                }
+    public int getPVRChannelGroupId(final String channelType, final String pvrChannelGroupName) {
+        List<KodiPVRChannelGroup> pvrChannelGroups = getPVRChannelGroups(channelType);
+        for (KodiPVRChannelGroup pvrChannelGroup : pvrChannelGroups) {
+            if (StringUtils.equalsIgnoreCase(pvrChannelGroup.getLabel(), pvrChannelGroupName)) {
+                return pvrChannelGroup.getId();
             }
         }
         return 0;
     }
 
-    private synchronized JsonArray getChannels(final int channelGroupID) {
-        JsonObject params = new JsonObject();
-        params.addProperty("channelgroupid", channelGroupID);
-        JsonElement response = socket.callMethod("PVR.GetChannels", params);
+    public synchronized List<KodiPVRChannel> getPVRChannels(final int pvrChannelGroupId) {
+        String method = "PVR.GetChannels";
+        String hash = method + "#channelgroupid=" + pvrChannelGroupId;
+        JsonElement response = REQUEST_CACHE.putIfAbsentAndGet(hash, () -> {
+            JsonObject params = new JsonObject();
+            params.addProperty("channelgroupid", pvrChannelGroupId);
+            return socket.callMethod(method, params);
+        });
 
+        List<KodiPVRChannel> pvrChannels = new ArrayList<>();
         if (response instanceof JsonObject) {
             JsonObject result = response.getAsJsonObject();
             if (result.has("channels")) {
-                return result.get("channels").getAsJsonArray();
+                for (JsonElement element : result.get("channels").getAsJsonArray()) {
+                    JsonObject object = (JsonObject) element;
+                    KodiPVRChannel pvrChannel = new KodiPVRChannel();
+                    pvrChannel.setId(object.get("channelid").getAsInt());
+                    pvrChannel.setLabel(object.get("label").getAsString());
+                    pvrChannel.setChannelGroupId(pvrChannelGroupId);
+                    pvrChannels.add(pvrChannel);
+                }
             }
         }
-        return null;
+        return pvrChannels;
     }
 
-    public int getChannelID(final int channelGroupID, final String channelName) {
-        JsonArray channels = getChannels(channelGroupID);
-        if (channels instanceof JsonArray) {
-            for (JsonElement element : channels) {
-                JsonObject channel = (JsonObject) element;
-                String label = channel.get("label").getAsString();
-                if (StringUtils.equalsIgnoreCase(label, channelName)) {
-                    return channel.get("channelid").getAsInt();
-                }
+    public int getPVRChannelId(final int pvrChannelGroupId, final String pvrChannelName) {
+        List<KodiPVRChannel> pvrChannels = getPVRChannels(pvrChannelGroupId);
+        for (KodiPVRChannel pvrChannel : pvrChannels) {
```

Comment:
I think that I would not assign this

Reply:
What do you mean exactly?
`for (KodiPVRChannel pvrChannel : getPVRChannels(pvrChannelGroupId)) {` instead?

Reply:
Yes